File: solutions/lua/binary-search/1/binary-search.lua
return function(array, target)
  local function binary_search(low, high)
    if low > high then return -1 end
    local avg = (low + high) // 2
    local value = array[avg]
    if target < value then return binary_search(low, avg - 1) end
    if target > value then return binary_search(avg + 1, high) end
    return avg
  end
  return binary_search(1, #array)
end

File: solutions/ruby/binary-search/1/binary_search.rb
class BinarySearch
  def initialize(array)= @array = array

  def search_for(target)
    binary(0, @array.size - 1, target)
  end

  private

  def binary(low, high, target)
    avg = (low + high) / 2
    if low > high then nil
    elsif target < @array[avg] then binary(low, avg - 1, target)
    elsif target > @array[avg] then binary(avg + 1, high, target)
    else avg
    end
  end
end
